File: scribbleshare-app/scribbleshare-frontend/src/scripts/canvas/canvasUpdate/CanvasUpdate.ts
import ByteBuffer from "../../protocol/ByteBuffer.js";
import {Canvas} from "../Canvas.js";
import CanvasUpdateType from "./CanvasUpdateType.js";
import CanvasUpdateInsert from "./canvasUpdates/CanvasUpdateInsert.js";
import CanvasUpdateMove from "./canvasUpdates/CanvasUpdateMove.js";
import CanvasUpdateDelete from "./canvasUpdates/CanvasUpdateDelete.js";

export default abstract class CanvasUpdate {
    dt: number;

    protected constructor(byteBuffer: ByteBuffer) {
        this.dt = byteBuffer.readUint8();
    }

    draw(canvas: Canvas, dt: number): boolean {
        return true;
    }

    abstract getType(): CanvasUpdateType;

    serialize(byteBuffer: ByteBuffer) {
        byteBuffer.writeUint8(this.getType());
    }

    static deserialize(type: CanvasUpdateType, byteBuffer: ByteBuffer): CanvasUpdate {
        let object: CanvasUpdate;
        switch (type) {
            case CanvasUpdateType.INSERT:
                object = new CanvasUpdateInsert(byteBuffer);
                break;
            case CanvasUpdateType.MOVE:
                object = new CanvasUpdateMove(byteBuffer);
                break;
            case CanvasUpdateType.DELETE:
                object = new CanvasUpdateDelete(byteBuffer);
                break;
            default:
                throw new Error('unknown canvasUpdateType ' + type);
        }
        return object;
    }
}

File: commons/src/main/java/net/stzups/scribbleshare/data/objects/canvas/canvasUpdate/CanvasUpdate.java
package net.stzups.scribbleshare.data.objects.canvas.canvasUpdate;

import io.netty.buffer.ByteBuf;
import net.stzups.scribbleshare.data.objects.canvas.Canvas;
import net.stzups.scribbleshare.data.objects.canvas.canvasUpdate.canvasUpdates.CanvasUpdateDelete;
import net.stzups.scribbleshare.data.objects.canvas.canvasUpdate.canvasUpdates.CanvasUpdateInsert;
import net.stzups.scribbleshare.data.objects.canvas.canvasUpdate.canvasUpdates.CanvasUpdateMove;
import net.stzups.scribbleshare.data.objects.exceptions.DeserializationException;
import net.stzups.scribbleshare.data.objects.exceptions.DeserializationTypeException;
import net.stzups.util.DebugString;

public abstract class CanvasUpdate {
    private final byte dt;

    public CanvasUpdate(ByteBuf byteBuf) {
        dt = byteBuf.readByte();
    }

    protected abstract CanvasUpdateType getCanvasUpdateType();

    public abstract void update(Canvas canvas, short id) throws CanvasUpdateException;

    public void serialize(ByteBuf byteBuf) {
        byteBuf.writeByte(dt);
    }

    @Override
    public String toString() {
        return DebugString.get(CanvasUpdate.class)
                .add("dt", dt)
                .toString();
    }

    static CanvasUpdate deserialize(CanvasUpdateType type, ByteBuf byteBuf) throws DeserializationException {
        CanvasUpdate canvasUpdate;
        switch (type) {
            case INSERT:
                canvasUpdate = new CanvasUpdateInsert(byteBuf);
                break;
            case MOVE:
                canvasUpdate = new CanvasUpdateMove(byteBuf);
                break;
            case DELETE:
                canvasUpdate = new CanvasUpdateDelete(byteBuf);
                break;
            default:
                throw new DeserializationTypeException(CanvasUpdateType.class, type);
        }
        return canvasUpdate;
    }
}
